sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -203,7 +206,7 @@ class Dataset[T] private[sql](
    * `fromRow` method later.
    */
   private val boundEnc =
-    exprEnc.resolveAndBind(logicalPlan.output, sparkSession.sessionState.analyzer)
+    exprEnc.resolveAndBind(planWithBarrier.output, sparkSession.sessionState.analyzer)
```

Comment:
I think we should only use `planWithBarrier` if necessary, this place is obviously unnecessary.